dask/distributed/worker.py

```diff
@@ -445,10 +461,20 @@ def collect(self, locations):
                 self.send_to_worker(worker, header, payload)
                 counter += 1
 
-            for i in range(counter):
-                queue.get()
+            msgs = [queue.get() for i in range(counter)]
+            for m in msgs:
+                if not m['got_key']:
+                    locations_copy[m['key']].remove(m['worker'])
+                    log(self.address, 'Failed to get key: ', m['key'],
+                        ' from worker: ', m['worker'])
+                else:
+                    locations_copy.pop(m['key'])
 
             del self.queues[qkey]
+            if locations_copy != {}:
+                log(self.address, 'Retrying collect with keys and locations',
+                    locations_copy)
+                self.collect(locations_copy)
```

Comment:
Oh nice, so you only need to recollect the missing keys.  

This might still run into problems if the data is genuinely missing.